docs/proposals/improve-addons.md

```diff
@@ -0,0 +1,350 @@
+```
+
+Note on CRDs in general: The AddonCatalogue strongly benefits from [OpenAPIv3's `oneOf` functionality](https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/), which is supported by Kubernetes. For example in the yaml above, you could say a source must be `oneOf` `docker` or `git`. The cool thing about `oneOf` is that `docker` and `git` can have totally have different fields from each other, but still would be statically validated. However, syntax like this is [not yet supported in kubebuilder](https://github.com/kubernetes-sigs/controller-tools/issues/461), which we use to generate the CRDs. Nonetheless, we would still propose to use kubebuilder for now as the advantages of an automated generation outweigh the downside of not being able to do oneOf (yet).
+
+### AddonController
+
+The AddonController resides in the seed cluster. Each seed cluster has its own instance of an AddonController. It has two main jobs:
+
+- selecting compatible addons from the AddonCatalogue for each user cluster
+- managing AddonInstallations for each user cluster. It is important to note that it only tracks which addon should be installed in which cluster, but does not do the installation of the addon itself (this task will be done by the AddonInstallation controller).
+
+The AddonController works by watching the AddonCatalogue CR from the Catalogue Distributor and managing an AddonCatalogue per UserCluster. It stores a version containing only compatible addons in the SeedCluster-namespace that corresponds to a UserCluster. In order to evaluate if an addon is compatible, the AddonController reads the `constraints` field of an AddonCatalogue CR.
+
+In order to track which addon should be installed in which cluster, the AddonController makes use of AddonInstallation CRs.
+
+### AddonInstallation CR
+
+- contains a mapping of a specific addon and cluster where to install
+  - the addon is specified by an addonRef. Currently, this consists of a name and version
+  - the cluster where to install is specified by the namespace the AddonInstallation was created in (`metadata.namespace`) →The AddonController makes sure the AddonInstallation CR is being placed in the correct namespace
+- tells if an add-on is a CoreAddon or not → We propose to have this per installation, as different user clusters can have different CoreAddons. Furthermore, it should be noted that the AddonController will always set the `coreAddon` field to false, while the CoreAddonController will always set it to true
+- contains the merged values that are being passed for the installation. This is needed as customers can set custom values for each installation.
+
+```yaml
+apiVersion: kubermatic.k8c.io/v1
+kind: AddonInstallation
+metadata:
+  name: prometheus
+    namespace: cluster-47s2ddlfgj
+spec:
+  coreAddon: false #This can be omitted, false will be the default
+    targetNamespace: monitoring
+    createNamespace: true # flag to create the targetNamespace
```

Comment:
this in't valid YAML. I'm not sure if it's intentional or not, but it's also happening in the example under `CoreAddonController`, so I wanted to point it out. Probably looking for:

```suggestion
spec:
  coreAddon: false #This can be omitted, false will be the default
  targetNamespace: monitoring
  createNamespace: true # flag to create the targetNamespace
```

Reply:
Good Spot! 
Yeah sometimes Nuclino Yaml hates you :D